### DISCUSSION TITLE
Global variable preservation between requests

### DISCUSSION THREAD

**Discussion by tinleym** [vercel/next.js#47605] [Help] [answered]
### Summary

I'm wondering how global variables are preserved between requests. If two different users fetch the component below, does the second user see the count as 2?

```
let count = 0
async function SomePage() {
   count++
   return <div>{count}</div>
}
```

**Comment by icyJoseph** [comment, marked as answer]
Uff asking the real questions here...

If the page is rendered at build time, then when a request arrives, "nothing" runs server side, unless revalidation kicks in, but that's an isolated process IIRC, so not triggered by a request, but by a sequence of requests.

> Note that during development mode you'd experience shared context anyway - 

```tsx
// in Dev Mode this snippet shares JavaScript context between incoming requests
// in Production Mode this page is pre-rendered at build time, and count is frozen - renders 1 for everybody
import styles from "./page.module.css";

let count = 0;

export default async function Home() {
  count++;
  return <main className={styles.main}>{count}</main>;
}
```

However, if you are doing on-demand generation, think GetServerSideProps or appDir pages that are "dynamic", then for sure there'll be a server side shared context. 

```tsx
import styles from "./page.module.css";

let count = 0;

// This page is not generated as the requests arrive
// each request mutates the count
export const dynamic = "force-dynamic";
// there's other ways to make a page dynamic, using cookies, headers, etc 

export default async function Home() {
  count++;
  return <main className={styles.main}>{count}</main>;
}
```

Often this is can also end up turning into a memory leak... like you might store data for a user in a globally shared object, and never clean it up... so it'll just stay there hogging  more and more memory over time.

However, serverless deployments are also a thing, and even though a context might be shared, it'll also go away when the task is killed by some kind of timeout. 

For example, on a Next.js Vercel deployment when a request arrives to GSSP, a task is spun if no other is present to take the request, and once things are idle, or if the task takes "too long", then the task is killed, and with it the JavaScript context.

As you can see, this is something to think about also on the `pages` directory. It is often one those pitfalls that we face when moving CSR apps to SSR/hybrid solutions. And it might also be an obscure issue because of serverless deployments.

**Comment by tinleym** [comment]
Thank you @icyJoseph! That's what I was afraid of. Since mutating server-side global variables does some combination of a) not updating when you expect them to b) resetting arbitrarily c) leaking memory and d) leaking user data, I thought that Next might have some special handling in lieu of an error or warning... especially to client-side devs adapting client-side code. I guess I'll suggest this!
